File: crates/sdk/src/artifact/protoc_gen_go_grpc.rs
use crate::{
    artifact::{language::go::GoBuilder, ArtifactSourceBuilder},
    context::ConfigContext,
};
use anyhow::Result;

pub async fn build(context: &mut ConfigContext) -> Result<String> {
    let name = "protoc-gen-go-grpc";

    let source_digest = "eba0f83ab252cffe2c6209f894c4c8238b2473a403bbdbcb985af25140aac95d";

    let source_version = "1.70.0";

    let source_path =
        format!("https://github.com/grpc/grpc-go/archive/refs/tags/v{source_version}.tar.gz");

    let source = ArtifactSourceBuilder::new(name, source_path.as_str())
        .with_digest(source_digest)
        .build();

    let build_dir = format!("grpc-go-{}/cmd/{}", source_version, name);

    GoBuilder::new(name)
        .with_build_dir(build_dir.as_str())
        .with_source(source)
        .build(context)
        .await
}

File: sdk/python/src/vorpal_sdk/artifact/protoc_gen_go_grpc.py
"""protoc-gen-go-grpc artifact (built from grpc-go source via the Go builder).

Mirrors ``sdk/typescript/src/artifact/protoc_gen_go_grpc.js``.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from vorpal_sdk.artifact import ArtifactSource
from vorpal_sdk.system import SYSTEMS

if TYPE_CHECKING:
    from vorpal_sdk.context import ConfigContext


class ProtocGenGoGrpc:
    """Builder for the protoc-gen-go-grpc artifact."""

    def build(self, context: ConfigContext) -> str:
        # Deferred import breaks the go-tool<->language-builder cycle.
        from vorpal_sdk.artifact.language.go import Go

        name = "protoc-gen-go-grpc"
        source_version = "1.79.1"
        source_path = (
            f"https://sdk.vorpal.build/source/"
            f"protoc-gen-go-grpc-v{source_version}.tar.gz"
        )

        source = ArtifactSource(name, source_path).build()

        build_directory = f"grpc-go-{source_version}/cmd/{name}"

        return (
            Go(name, SYSTEMS)
            .with_aliases([f"{name}:{source_version}"])
            .with_build_directory(build_directory)
            .with_source(source)
            .build(context)
        )
